### PR TITLE
Dshank/position loss

### PR DESCRIPTION
**Pull request by Concomitant**
If anyone sees some weirdness, I'm all ears.

loss swings around wildly, am thinking it might make sense to just predict either just top left position, top left position + offset, or the centroid of the box. That way it doesn't need to guess the dimensions of tokens independently so much.

### CODE REVIEW

**Review by lhz1029 — COMMENTED** [review]
So it looks like you're masking a random subset of the token positions, but predicting the positions for all the tokens. And at some point you were also experimenting with masking out text and position for some tokens, but at least in the current work, it's just been position, i.e. all text is in tact.

Have all these combinations caused similar stability issues? I'd be curious to know what the outcomes of these slight variations has been.

**Comment by Concomitant** [comment]
The tasks seemed a bit harder, but as people have pointed out, the training regime was always bad. I think priority is getting that more or less fixed. I also would like to be able to compare the results of those experiments!

File: finetune/nn/target_blocks.py [hunk 1]
```diff
@@ -65,9 +65,76 @@ def cps_model(M, cps_mask, hidden, config):
...
+            "output_bias_posreg",
+            #shape=[n_vocab],
+            # We are regressing into 4D space...
+            # might be a good idea to do top left + offset?
+            shape=[4],
+            initializer=tf.zeros_initializer()
+        )
+        
+        # Going to add a new variable to project downwards
+        regression_matrix = tf.get_variable(
```
Comment (madisonmay) [inline finetune/nn/target_blocks.py:109]: I feel like rather than projecting out to n_vocab and projecting down you should just directly project down to the shape of the position outputs.

File: finetune/nn/target_blocks.py [hunk 2]
```diff
@@ -65,9 +65,76 @@ def cps_model(M, cps_mask, hidden, config):
...
+        )
+
+        logits = tf.matmul(normed_proj, embed_weights, transpose_b=True)
+        logits = tf.reshape(tf.matmul(logits, regression_matrix), [-1,4])
+        logits = tf.nn.bias_add(logits, output_bias)
+        # Swapping out old logit stuff
+        #mlm_loss = tf.contrib.losses.sparse_softmax_cross_entropy(
+        mlm_loss = tf.contrib.losses.mean_squared_error(
+            #logits,
+            tf.broadcast_to(output_bias, tf.shape(targets)),
```
Comment (madisonmay) [inline finetune/nn/target_blocks.py:123]: Only comment here is that I don't think you get a lot of value out of running this experiment that you wouldn't get out of purely computing the mean and calculating MSE offline. The point in my mind was just to get some sort of baseline that we could use to understand how well the model is performing as we didn't even have a qualitative sense of what range of losses was reasonable.

As I think about this more, when you continue these experiments it would be nice to log something like mean error per dim or something that would give us a quick way to get a qualitative sense of how well the model is doing.


I.e.:

```
mean_error = tf.mean(tf.math.abs((logits - targets), axis=0)
tf.summary.scalar("Top", mean_error[0])
tf.summary.scalar("Bottom", mean_error[1])
...
```

File: finetune/nn/target_blocks.py (hunk 1 above)
Comment (Concomitant) [reply finetune/nn/target_blocks.py:109]: Yeah, I suppose it isn't like I'm really preserving anything in the logits, so that seems like a change to try. My thought was that I wanted to preserve as much as the structure as possible but now that I think about it I doubt it's that relevant.

File: finetune/nn/target_blocks.py (hunk 2 above)
Comment (Concomitant) [reply finetune/nn/target_blocks.py:123]: That seems like a reasonable thing to do. I might actually want to try absolute error, also!

Comment (madisonmay) [reply finetune/nn/target_blocks.py:123]: yeah i think only argument against MAE is that you have a sharp discontinuity at 0, but that's only relevant late in training when you're trying to eek out the last bit of performance

File: finetune/nn/target_blocks.py (hunk 1 above)
Comment (madisonmay) [reply finetune/nn/target_blocks.py:109]: There are decent reasons why you don't want this -- mainly the final layers of model are the ones that are most task specific that don't transfer as well. This is why you lop off the MLM head of bert at finetuning time -- those params aren't well suited to helping solve a classification task, for instance.

**Comment by madisonmay** [comment]
Closing as this is lower priority for now
